validator/delphi_validator/validate.py

```diff
@@ -1,237 +1,721 @@
+        # Run checks for recent dates in each geo-sig combo vs semirecent (last week) API data.
+        for geo_sig_df, geo, sig in read_geo_sig_cmbo_files(
+                geo_sig_cmbo,
+                export_dir,
+                [name_match_pair[0] for name_match_pair in validate_files],
+                date_slist):
+
+            match_obj = smooth_option_regex.match(sig)
+            smooth_option = match_obj.group(1)
+
+            if smooth_option not in ('raw', 'smoothed'):
+                smooth_option = 'smoothed' if '7dav' in sig or 'smoothed' in sig else 'raw'
+
+            weight_option = 'weighted' if 'wili' in sig or 'wcli' in sig else 'unweighted'
+
+            print("Printing geo_sig_df scenes:", geo_sig_df.shape)
+            print(geo_sig_df)
+
+            max_date = geo_sig_df["time_value"].max()
+            self.check_min_allowed_max_date(max_date, weight_option, geo, sig)
+            self.check_max_allowed_max_date(max_date, geo, sig)
+
+            # Check data from a group of dates against recent (previous 7 days, by default)
+            # data from the API.
+            for checking_date in date_list:
+                recent_cutoff_date = checking_date - recent_lookbehind
+                recent_df = geo_sig_df.query(
+                    'time_value <= @checking_date & time_value >= @recent_cutoff_date')
+
+                # Reference dataframe runs backwards from the checking_date
+                reference_start_date = checking_date - \
+                    min(semirecent_lookbehind, self.max_check_lookbehind)
+                reference_end_date = recent_cutoff_date - timedelta(days=1)
+                reference_api_df = fetch_api_reference(
+                    self.data_source, reference_start_date, reference_end_date, geo, sig)
+
+                self.check_max_date_vs_reference(
+                    recent_df, reference_api_df, checking_date, geo, sig)
+
+                if self.sanity_check_rows_per_day:
+                    self.check_rapid_change(
+                        recent_df, reference_api_df, checking_date, geo, sig)
+
+                if self.sanity_check_value_diffs:
+                    self.check_avg_val_diffs(
+                        recent_df, reference_api_df, smooth_option, checking_date, geo, sig)
+
+            # Keeps script from checking all files in a test run.
+            if self.test_mode:
+                kroc += 1
+                if kroc == 2:
+                    break
+
+        self.exit()
+
+    def exit(self):
+        """
+        If any not-suppressed exceptions were raised, print and exit with non-zero status.
+        """
+        if self.raised_errors:
+            suppressed_counter = 0
+            subset_raised_errors = []
+
+            for val_error in self.raised_errors:
+                raised_check_id = tuple(item.strftime("%Y-%m-%d") if isinstance(
+                    item, (date, datetime)) else item for item in val_error.check_data_id)
```

Comment:
Add parentheses to clarify whether this is `(str, <generator>)` or `(str, str, ..., str)` -- operator precedence in this context varies across languages